src/Common/ThreadProfileEvents.cpp

```diff
@@ -104,6 +117,387 @@ void TasksStatsCounters::incrementProfileEvents(const ::taskstats & prev, const
+
+    // check permissions
+    // cat /proc/sys/kernel/perf_event_paranoid
+    // -1: Allow use of (almost) all events by all users
+    // >=0: Disallow raw tracepoint access by users without CAP_IOC_LOCK
+    // >=1: Disallow CPU event access by users without CAP_SYS_ADMIN
+    // >=2: Disallow kernel profiling by users without CAP_SYS_ADMIN
+    // >=3: Disallow all event access by users without CAP_SYS_ADMIN
+    Int32 perf_event_paranoid = 0;
+    std::ifstream paranoid_file("/proc/sys/kernel/perf_event_paranoid");
+    paranoid_file >> perf_event_paranoid;
+
+    bool has_cap_sys_admin = hasLinuxCapability(CAP_SYS_ADMIN);
+    if (perf_event_paranoid >= 3 && !has_cap_sys_admin)
+    {
+        LOG_WARNING(&Poco::Logger::get("PerfEventsCounters"), "Not enough permissions to record perf events: "
+            "perf_event_paranoid = {} and CAP_SYS_ADMIN = 0",
+            perf_event_paranoid);
+        return false;
+    }
+
+    // check file descriptors limit
+    rlimit64 limits{};
+    if (getrlimit64(RLIMIT_NOFILE, &limits))
+    {
+        LOG_WARNING(&Poco::Logger::get("PerfEventsCounters"), "Unable to get rlimit: {} ({})", strerror(errno),
+                    errno);
+        return false;
+    }
+    UInt64 maximum_open_descriptors = limits.rlim_cur;
+
+    const size_t opened_descriptors = std::distance(
+        std::filesystem::directory_iterator("/proc/self/fd"),
+        std::filesystem::directory_iterator());
+
+    UInt64 fd_count_afterwards = opened_descriptors + events_to_open.size();
+    UInt64 threshold = static_cast<UInt64>(maximum_open_descriptors * FILE_DESCRIPTORS_THRESHOLD);
+    if (fd_count_afterwards > threshold)
+    {
+        LOG_WARNING(&Poco::Logger::get("PerfEventsCounters"), "Can't measure perf events as the result number of file descriptors ({}) is more than the current threshold ({} = {} * {})",
+            fd_count_afterwards, threshold, maximum_open_descriptors,
+            FILE_DESCRIPTORS_THRESHOLD);
+        return false;
+    }
+
+    // open descriptors for new events
+    for (size_t i : events_to_open)
+    {
+        const PerfEventInfo & event_info = raw_events_info[i];
+        int & fd = thread_events_descriptors_holder.descriptors[i];
+        // disable by default to add as little extra time as possible
+        fd = openPerfEventDisabled(perf_event_paranoid, has_cap_sys_admin, event_info.event_type, event_info.event_config);
+
+        if (fd == -1)
+        {
+            LOG_WARNING(&Poco::Logger::get("PerfEventsCounters"), "Perf event is unsupported: {}"
+                " (event_type={}, event_config={})",
+                event_info.settings_name, event_info.event_type,
+                event_info.event_config);
+        }
+    }
+
+    return true;
+}
+
+// Parse comma-separated list of event names. Empty or 'all' means all available
+// events.
+// TODO add validation to setting
+std::vector<size_t> PerfEventsCounters::eventIndicesFromString(const std::string & events_list)
```

Comment:
Comma separated list will be processed incorrectly if there are spaces around event names. For example, "cpu, instructions" will be handled as "cpu" and " instructions" events where " instructions" will not count as it has a space in the beginning. Seems important for me as comma is followed by a space in the most of the languages I know